freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/update/action/UserDataUpdateActions.java

```diff
@@ -0,0 +1,124 @@
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(StackContext context) {
+                return new UserDataUpdateRequest(context.getStack().getId());
+            }
+
+            @Override
+            protected Object getFailurePayload(UserDataUpdateRequest payload, Optional<StackContext> flowContext, Exception ex) {
+                return new UserDataUpdateFailed(payload.getResourceId(), ex);
+            }
+        };
+    }
+
+    @Bean(name = "UPDATE_USERDATA_ON_PROVIDER_STATE")
+    public AbstractUserDataUpdateAction<?> updateUserDataOnProviderSide() {
+        return new AbstractUserDataUpdateAction<>(UserDataUpdateSuccess.class) {
+            @Override
+            protected void doExecute(StackContext context, UserDataUpdateSuccess payload, Map<Object, Object> variables) {
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(StackContext context) {
+                Stack stack = context.getStack();
+                String userData = stack.getImage().getUserdata();
+                List<CloudResource> cloudResources = getCloudResources(stack.getId());
+                return new UserDataUpdateOnProviderRequest(context.getCloudContext(), context.getCloudCredential(), context.getCloudStack(),
+                        cloudResources, userData);
+            }
+
+            @Override
+            protected Object getFailurePayload(UserDataUpdateSuccess payload, Optional<StackContext> flowContext, Exception ex) {
+                return new UserDataUpdateFailed(UPDATE_USERDATA_FAILED_EVENT.event(), payload.getResourceId(), ex);
+            }
+        };
+    }
+
+    @Bean(name = "UPDATE_USERDATA_FINISHED_STATE")
+    public AbstractUserDataUpdateAction<?> finishUserDataUpdate() {
+        return new AbstractUserDataUpdateAction<>(UserDataUpdateOnProviderResult.class) {
+            @Override
+            protected void doExecute(StackContext context, UserDataUpdateOnProviderResult payload, Map<Object, Object> variables) {
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(StackContext context) {
+                return new StackEvent(UpdateUserDataEvents.UPDATE_USERDATA_FINISHED_EVENT.selector(), context.getStack().getId());
+            }
+
+            @Override
+            protected Object getFailurePayload(UserDataUpdateOnProviderResult payload, Optional<StackContext> flowContext, Exception ex) {
+                return new UserDataUpdateFailed(UPDATE_USERDATA_FAILED_EVENT.event(), payload.getResourceId(), ex);
+            }
+        };
+    }
+
+    @Bean(name = "UPDATE_USERDATA_FAILED_STATE")
+    public Action<?, ?> handleUpdateUserDataFailure() {
+        return new UserDataUpdateFailureHandlerAction();
+    }
+
+    private List<CloudResource> getCloudResources(Long stackId) {
```

Comment:
as this is used only in `updateUserDataOnProviderSide` I think it should be in that anonymous class